### PR TITLE
Issue 582

### LINKED ISSUE #582: Test Ids Are Not Unique
When using multiple processes or multiple domains, the test ids are not unique.  This is an issue for identifying unique tests from the GUI and in the TC service messages. 

One possible solution would be to prefix the id with some kind of process/domain id. 

For example using the default single process, separate app domain you could get the following pattern
`p1-d1-testId`

Running `process=Multiple` or `process=Parallel` you might see the following
`p1-d1-testId`
`p2-d1-testId`

If you used a mixture of project files and assemblies, where the project creates additional processes/domains, for 1 project w/ one assembly and two addition assemblies you might see
`p1-d1-testId` // project created additional process and domain
`p2-d1-testId` // assembly in its own process also has a d1

Another solution could be to prefix the test id with a identifier for the assembly that the test resides in. A hash of the assembly or a guid could be potentially be used here. This solution should require less coordination between processes/domains.

### PR DESCRIPTION
**Pull request by CharliePoole**
**Now ready to merge. See comment below.**

Fixes #582 

This could be merged harmlessly, but there doesn't seem to be a need right now. However, if anyone is going to work in areas that are affected, feel free to merge it.

The current state of the code is that work on the framework side is done. A new static property of the Test class contains a prefix that is used for all ids. The framework controller sets the id based on an argument passed in from the driver.

Currently, the driver is hard-coded to pass in the string "ID" so the only change is that test ids are of the form "IDxxxx" rather than just xxxx.

The v2 driver now generates ids of the form rrrr-tttt, where rrrr is the runner id and tttt is the test id. However, the framework is currently always passed a runner id of 1, giving us ids like 1-xxxx. This will eventually be changed, once the engine is generating IDs.

Bottom line: the framework works, the ids are all prefixed but they are not yet unique. 

Next steps: Work in the engine to create unique ids for each assembly.

### CODE REVIEW

**Comment by agray** [comment]
Test comment only.

**Comment by CharliePoole** [comment]
@agray Thanks, so it looks like anyone who wants to can comment on these. That's good.

**Comment by CharliePoole** [comment]
With this last commit, it should be ready to merge.

TestPackage has been substantially modified, which is a far-reaching change. It now contains subpackages as needed. The top level package is an anonymous package based on what was enterered on the command-line. Each SubPackage represents one file name entered. Those files which are actually projects (.nunit, .csproj, etc.) will have further subpackages. Every package has a unique id in the form of string representing an integer value.

Package ids become driver ids when we create drivers for the package. Each driver uses the id in some way to guarantee that test ids are unique. Currently, all three drivers create test ids in the form ddd-ttt where ddd is the driver id and ttt is the unique test identifier. This is done in the framework itself for NUnit 3 and in the driver for the other two drivers we currently support.

This change really only impacts what is shown in the XML file as a test id but it enables other changes to be made, which require a unique id across all assemblies.

File: src/NUnitEngine/nunit.engine.api/TestPackage.cs [hunk 1]
```diff
@@ -63,8 +61,19 @@ public TestPackage(string filePath)
...
             foreach (string testFile in testFiles)
-                Add(Path.GetFullPath(testFile));
+                SubPackages.Add(new TestPackage(testFile));
+        }
+
+        private static int _nextID = 0;
+
+        private string GetNextID()
+        {
+            return unchecked(_nextID++).ToString();
```
Comment (rprouse) [inline src/NUnitEngine/nunit.engine.api/TestPackage.cs]: For my own education, why did you use unchecked here? I thought unchecked was to prevent compile time overflow checks with constant values and that "Expressions that contain non-constant terms are unchecked by default at compile time and run time."

Do we have the /checked compiler option enabled?

Comment (CharliePoole) [inline src/NUnitEngine/nunit.engine.api/TestPackage.cs]: Actually, it's carried over from some very old code that started out in .NET 1.0. It's been copied forward for years and I duplicated it here.  It was used for numbering tests and the idea is that we don't want an excess of  tests to cause the whole run to fail, even if it results in duplicate ids. IIRC, 'unchecked' was needed for some compilers but not others.

In reality, it's not needed here. We might have more than MAXINT tests but we aren't going to have that many packages in a run. Probably I should remove it since it's potentially confusing.

File: src/NUnitEngine/nunit.engine.api/TestPackage.cs [hunk 2]
```diff
@@ -86,32 +104,49 @@ public string Name
...
+        /// <param name="name">The name of the setting</param>
+        /// <param name="value">The value of the setting</param>
+        /// <remarks>
+        /// Once a package is created, subpackages may have been created
+        /// as well. If you add a setting directly to the Settings dictionary
+        /// of the package, the subpackages are not updated. This method is
+        /// used when the settings are intended to be reflected to all the
+        /// subpackages under the package.
+        /// </remarks>
+        public void AddSetting<T>(string name, T value)
```
Comment (rprouse) [inline src/NUnitEngine/nunit.engine.api/TestPackage.cs]: Does this really need to be generic? The settings are collections of `object`, so why not just pass in an `object` for the `value`?

File: src/NUnitFramework/framework/Api/FrameworkController.cs [hunk 3]
```diff
@@ -58,11 +58,12 @@ public class FrameworkController : MarshalByRefObject
         /// <summary>
         /// Construct a FrameworkController using the default builder and runner.
         /// </summary>
-        public FrameworkController(string assemblyPath, IDictionary settings)
+        public FrameworkController(string assemblyPath, string idPrefix, IDictionary settings)
```
Comment (rprouse) [inline src/NUnitFramework/framework/Api/FrameworkController.cs]: It wasn't done before, but this is part of the public API, so the parameters should be in the XMLDOC.

**Comment by rprouse** [comment]
Looks good. A few minor points. Make changes if you want and merge away :+1:

File: src/NUnitEngine/nunit.engine.api/TestPackage.cs (hunk 2 above)
Comment (CharliePoole) [inline src/NUnitEngine/nunit.engine.api/TestPackage.cs]: You're right... we have various settings collections, but this one is quite primitive and doesn't need to be generic at all.

File: src/NUnitFramework/framework/Api/FrameworkController.cs (hunk 3 above)
Comment (CharliePoole) [inline src/NUnitFramework/framework/Api/FrameworkController.cs]: Will do.

File: src/NUnitEngine/nunit.engine.api/TestPackage.cs (hunk 1 above)
Comment (CharliePoole) [inline src/NUnitEngine/nunit.engine.api/TestPackage.cs]: I removed unchecked.
